pageserver/src/storage_sync.rs

```diff
@@ -1347,20 +1395,42 @@ fn compare_local_and_remote_timeline(
     new_sync_tasks: &mut VecDeque<(TenantTimelineId, SyncTask)>,
     sync_id: TenantTimelineId,
     local_metadata: TimelineMetadata,
-    local_files: HashSet<PathBuf>,
+    local_files: HashMap<PathBuf, LayerFileMetadata>,
     remote_entry: &RemoteTimeline,
 ) -> (LocalTimelineInitStatus, bool) {
     let _entered = info_span!("compare_local_and_remote_timeline", sync_id = %sync_id).entered();
 
-    let remote_files = remote_entry.stored_files();
+    let needed_to_download_files = remote_entry
+        .stored_files()
+        .iter()
+        .filter_map(|(layer_file, remote_metadata)| {
+            if let Some(local_metadata) = local_files.get(layer_file) {
+                match (remote_metadata.file_size(), local_metadata.file_size()) {
+                    (Some(x), Some(y)) if x == y => { None },
+                    _ => {
```

Comment:
Test sounds like a good idea, I had missed that `(None, Some(local_layer_size))`.

Reply:
At the moment this would just log:

```
2022-10-14T14:05:01.785967Z  WARN init_tenant_mgr:compare_local_and_remote_timeline{sync_id=deb8658155a6c7e2e0c7dac0b4f09c6d/b01ef4ffbdddd6c774c06d8da4592666}: Redownloading locally existing "$REPO/tenants/deb8658155a6c7e2e0c7dac0b4f09c6d/timelines/b01ef4ffbdddd6c774c06d8da4592666/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000000169ACF0-000000000169AD69" due to size mismatch, size on index: None, on disk: Some(25600000)
```

So it's very good find.

Reply:
Fixed and test added in f2b08f19.